sensor/common/compliance/nodescan_handler_impl.go

```diff
@@ -1,36 +1,50 @@
 package compliance
 
 import (
+	"sync/atomic"
+
 	"github.com/pkg/errors"
 	"github.com/stackrox/rox/generated/internalapi/central"
 	"github.com/stackrox/rox/generated/storage"
 	"github.com/stackrox/rox/pkg/centralsensor"
 	"github.com/stackrox/rox/pkg/concurrency"
+	"github.com/stackrox/rox/pkg/sync"
 )
 
 type nodeScanHandlerImpl struct {
 	nodeScans <-chan *storage.NodeScanV2
-	toCentral chan *central.MsgFromSensor
+	toCentral <-chan *central.MsgFromSensor
 
-	stopC    concurrency.ErrorSignal
+	// lock prevents the race condition between Start() [writer] and ResponsesC() [reader]
+	lock      *sync.Mutex
+	numStarts uint32
+	stopC     concurrency.ErrorSignal
+	// stoppedC is signaled when the goroutine inside of run() finishes
 	stoppedC concurrency.ErrorSignal
 }
 
+func (c *nodeScanHandlerImpl) Stopped() concurrency.ReadOnlyErrorSignal {
+	return &c.stoppedC
+}
+
 func (c *nodeScanHandlerImpl) Capabilities() []centralsensor.SensorCapability {
 	return []centralsensor.SensorCapability{centralsensor.NodeScanningCap}
 }
 
 func (c *nodeScanHandlerImpl) ResponsesC() <-chan *central.MsgFromSensor {
+	c.lock.Lock()
+	defer c.lock.Unlock()
 	return c.toCentral
 }
 
 func (c *nodeScanHandlerImpl) Start() error {
-	// c.run closes chan toCentral on exit, so we do not want to close twice in case of a restart
-	if !c.stopC.IsDone() {
-		go c.run()
-		return nil
+	if !atomic.CompareAndSwapUint32(&c.numStarts, 0, 1) {
+		return errors.New("unable to start - component reached the maximum number of starts: 1")
```

Comment:
I suggest to make the error message a bit more human-friendly

```suggestion
		return errors.New("unable to start the component more than once")
```

Reply:
Got it covered in 0cdd69cd0e1ad6887951afad8bea2d756a4c8565